android/src/main/java/com/twiliorn/library/PatchedVideoView.java

```diff
@@ -45,6 +64,108 @@ public void run() {
             });
         }
         super.onFrame(frame);
+
+        if (captureThisFrame) {
+            captureThisFrame = false;
+            saveVideoFrame(frame);
+        }
+    }
+
+    /**
+     * Create an NV21Buffer with the same pixel content as the given I420 buffer. Ripped from here: https://stackoverflow.com/a/55158228
+     */
+    private byte[] createNV21Data(VideoFrame.I420Buffer i420Buffer) {
+        final int width = i420Buffer.getWidth();
+        final int height = i420Buffer.getHeight();
+        final int chromaStride = width;
+        final int chromaWidth = (width + 1) / 2;
+        final int chromaHeight = (height + 1) / 2;
+        final int ySize = width * height;
+        final ByteBuffer nv21Buffer = ByteBuffer.allocateDirect(ySize + chromaStride * chromaHeight);
+        // We don't care what the array offset is since we only want an array that is direct.
+        @SuppressWarnings("ByteBufferBackingArray") final byte[] nv21Data = nv21Buffer.array();
+        for (int y = 0; y < height; ++y) {
+            for (int x = 0; x < width; ++x) {
+                final byte yValue = i420Buffer.getDataY().get(y * i420Buffer.getStrideY() + x);
+                nv21Data[y * width + x] = yValue;
+            }
+        }
+        for (int y = 0; y < chromaHeight; ++y) {
+            for (int x = 0; x < chromaWidth; ++x) {
+                final byte uValue = i420Buffer.getDataU().get(y * i420Buffer.getStrideU() + x);
+                final byte vValue = i420Buffer.getDataV().get(y * i420Buffer.getStrideV() + x);
+                nv21Data[ySize + y * chromaStride + 2 * x + 0] = vValue;
+                nv21Data[ySize + y * chromaStride + 2 * x + 1] = uValue;
+            }
+        }
+        return nv21Data;
+    }
+
+    private void sendEvent(String event, WritableMap params) {
+        ((ReactContext) this.getContext()).getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
+                .emit(event, params);
+    }
+
+    private void saveVideoFrame(VideoFrame frame) {
+        frame.retain();
+        Context context = this.getContext();
+        String fileId = UUID.randomUUID().toString();
+        String filename = "rntframe-" + fileId + ".jpeg";
+        VideoFrame.I420Buffer i420Buffer = frame.getBuffer().toI420();
+        final int width = i420Buffer.getWidth();
+        final int height = i420Buffer.getHeight();
+        byte[] nv21Data = createNV21Data(i420Buffer);
+        YuvImage yuvImage = new YuvImage(nv21Data, ImageFormat.NV21, width, height, null);
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        yuvImage.compressToJpeg(new Rect(0, 0, width, height), 100, out);
```

Comment:
🔧 `saveVideoFrame()` blocks further video rendering until it completes. Which means any operations that might take time should be handled asynchronously so that the user doesn't experience video stutter when Streemshots are being captured.

Once you've converted `frame` into `nv21Data`, the frame is no longer needed. So at this point you can return from this function to allow video rendering to continue, and do the remaining heavy lifting (JPEG compression, file I/O) in a background thread. Let me know if you're not familiar with the best way to do that in Android native code and I can help out. Unfortunately since this isn't Kotlin you can't use the friendlier Kotlin coroutine approach and will have to use one of the older Android/Java solutions for async processing.